### PR TITLE
LibJS: Implement full destructuring support

### PR DESCRIPTION
**Pull request by mattco98**
This PR fully (as far as I know, hence the draft) implements destructuring support. This includes proper array elision handling, as well as arbitrarily-nested destructuring clauses. It works in both assignment context and function parameter context. The spec seems to differentiate these into `BindingProperty`'s and `AssignmentProperty`'s, but they're similar enough that they can be lumped together. 

For your viewing pleasure, take a look at this monstrosity:
```js
> let { a: { b: [,, { d, e = 40 }] } } = { a: { b: [10, 20, { d: 30 }] } };
undefined
> d
30
> e
40
```

Currently this gives us ~400 new passing tests. 

TODO:
- [x] Verify there are no regressions and check test results for unsupported edge cases
- [x] Implement support in the bytecode generator

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/VM.cpp [hunk 1]
```diff
@@ -207,18 +207,15 @@ void VM::assign(const NonnullRefPtr<BindingPattern>& target, Value value, Global
...
+        for (size_t i = 0; i < binding.properties.size(); i++) {
             if (exception())
                 return;
 
-            if (index >= binding.properties.size())
-                break;
-
-            auto pattern_property = binding.properties[index];
-            ++index;
+            auto pattern_property = binding.properties[i];
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/VM.cpp:214]: Let's not copy this, it could involve three different refcount increment/decrements

File: Userland/Libraries/LibJS/Runtime/VM.cpp [hunk 2]
```diff
@@ -207,18 +207,15 @@ void VM::assign(const NonnullRefPtr<BindingPattern>& target, Value value, Global
...
 
             if (pattern_property.is_rest) {
+                VERIFY(i == binding.properties.size());
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/VM.cpp:217]: This check seems impossible, you're verifying that the element you just grabbed is one past the end?

File: Userland/Libraries/LibJS/Runtime/VM.cpp [hunk 3]
```diff
@@ -262,52 +263,60 @@ void VM::assign(const NonnullRefPtr<BindingPattern>& target, Value value, Global
...
                 return;
 
-            if (pattern_property.name) {
-                set_variable(pattern_property.name->string(), value, global_object, first_assignment, specific_scope);
-                if (pattern_property.is_rest)
-                    break;
-                continue;
-            }
+            auto& alias = pattern_property.alias;
+            if (!alias.has<Empty>()) {
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/VM.cpp:267]: This chain of ifs will check the index multiple times, and is a bit too verbose
Could use `visit` to match the type instead
```c++
alias.visit(
    [&](Empty) {},
    [&](NonnullRefPtr<Identifier> const& ident) { set_variable(ident->string(), ...); },
    [&] (BindingPattern const& pattern) { assign(pattern, ...); });
```

File: Userland/Libraries/LibJS/Runtime/VM.cpp [hunk 4]
```diff
@@ -262,52 +263,60 @@ void VM::assign(const NonnullRefPtr<BindingPattern>& target, Value value, Global
...
             } else {
-                value_to_assign = object->get(property.name->string());
+                if (property.name.has<NonnullRefPtr<Identifier>>()) {
+                    assignment_name = property.name.get<NonnullRefPtr<Identifier>>()->string();
+                } else {
+                    auto expression = property.name.get<NonnullRefPtr<Expression>>()->execute(interpreter(), global_object);
+                    assignment_name = expression.to_property_key(global_object);
+                    if (exception())
+                        break;
+                }
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/VM.cpp:315]: This if can also use .visit, though it's not as bad as the one before :P

File: Userland/Libraries/LibJS/Runtime/VM.cpp (hunk 2 above)
Comment (mattco98) [reply Userland/Libraries/LibJS/Runtime/VM.cpp]: It's left over from a previous iteration of the code. This is a draft, after all -- I wasn't expecting a review so soon :)

Comment (alimpfard) [reply Userland/Libraries/LibJS/Runtime/VM.cpp]: Me: sees an interesting draft
Me: ignores "draft"

**Comment by mattco98** [comment]
Since I've verified there are no regressions, it's time for review :)

Test262 test result diff:
img

File: Userland/Libraries/LibJS/Bytecode/Op.h [hunk 5]
```diff
@@ -49,6 +49,21 @@ class LoadImmediate final : public Instruction {
     Value m_value;
 };
 
+class LoadString final : public Instruction {
```
Comment (awesomekling) [inline Userland/Libraries/LibJS/Bytecode/Op.h:52]: This already exists! It's called `NewString` :^)

Comment (mattco98) [reply Userland/Libraries/LibJS/Bytecode/Op.h:52]: Oops!

**Comment by alimpfard** [comment]
Sorry for the conflicts ^^'

File: Userland/Libraries/LibJS/Bytecode/Op.h [hunk 6]
```diff
@@ -377,6 +377,17 @@ class JumpNullish final : public Jump {
...
+public:
+    explicit JumpUndefined(Optional<Label> true_target = {}, Optional<Label> false_target = {})
+        : Jump(Type::JumpUndefined, move(true_target), move(false_target))
+    {
+    }
+
+    void execute_impl(Bytecode::Interpreter&) const;
+    String to_string_impl(Bytecode::Executable const&) const;
+};
+
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Bytecode/Op.h:432]: Please also add this new jump type to `Passes/GenerateCFG.cpp` line 61.

File: Userland/Libraries/LibJS/Bytecode/Op.cpp [hunk 7]
```diff
@@ -418,7 +452,16 @@ String NewArray::to_string_impl(Bytecode::Executable const&) const
     return builder.to_string();
 }
 
+<<<<<<< HEAD
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Bytecode/Op.cpp]: :eyes:

Comment (mattco98) [reply Userland/Libraries/LibJS/Bytecode/Op.cpp]: I blame your merge conflicts!
